lang/src/org/partiql/lang/ast/passes/StatementRedactor.kt

```diff
@@ -82,238 +73,249 @@ fun redact(statement: String,
+        when (args.size) {
+            1 -> redactExpr(args[0])
+            2 -> {
+                redactExpr(args[0])
+                redactExpr(args[1])
+            }
+            else -> throw IllegalArgumentException(INVALID_NUM_ARGS)
+        }
     }
 
-    private fun redactOnSeq(seq: Seq) {
-        seq.values.map {
-            redactOnField(it)
+    private fun arithmeticOpRedaction(args: List<PartiqlAst.Expr>) {
+        if (args.size != 2) {
+            throw IllegalArgumentException(INVALID_NUM_ARGS)
         }
+        redactExpr(args[0])
+        redactExpr(args[1])
     }
 
-    private fun redactOnStruct(struct: Struct) {
-        struct.fields.map {
-            if (it.name is Literal) {
-                redactOnLiteral(it.name)
+    private fun redactNAry(node: PartiqlAst.Expr) {
+        when (node) {
+            // Logical Ops
+            is PartiqlAst.Expr.And -> redactLogicalOp(node.operands)
+            is PartiqlAst.Expr.Or -> redactLogicalOp(node.operands)
+            is PartiqlAst.Expr.Not -> redactExpr(node.expr)
+            // Comparison Ops
+            is PartiqlAst.Expr.Eq -> redactComparisonOp(node.operands)
+            is PartiqlAst.Expr.Ne -> redactComparisonOp(node.operands)
+            is PartiqlAst.Expr.Gt -> redactComparisonOp(node.operands)
+            is PartiqlAst.Expr.Gte -> redactComparisonOp(node.operands)
+            is PartiqlAst.Expr.Lt -> redactComparisonOp(node.operands)
+            is PartiqlAst.Expr.Lte -> redactComparisonOp(node.operands)
+            is PartiqlAst.Expr.InCollection -> redactComparisonOp(node.operands)
+            // Arithmetic Ops
+            is PartiqlAst.Expr.Plus -> plusMinusRedaction(node.operands)
+            is PartiqlAst.Expr.Minus -> plusMinusRedaction(node.operands)
+            is PartiqlAst.Expr.Times -> arithmeticOpRedaction(node.operands)
+            is PartiqlAst.Expr.Divide -> arithmeticOpRedaction(node.operands)
+            is PartiqlAst.Expr.Modulo -> arithmeticOpRedaction(node.operands)
+            is PartiqlAst.Expr.Concat -> arithmeticOpRedaction(node.operands)
+            // BETWEEN
+            is PartiqlAst.Expr.Between -> {
+                if (!skipRedaction(node.value, safeFieldNames)) {
+                    redactExpr(node.from)
+                    redactExpr(node.to)
+                }
             }
-            redactOnField(it.expr)
+            // CALL
+            is PartiqlAst.Expr.Call -> redactCall(node)
+            else -> { /* intentionally blank */ }
         }
     }
 
-    private fun redactOnCall(args: List<ExprNode>) {
-        if (args.isEmpty()) {
-            return
-        }
-        if (args[0] !is VariableReference) {
-            return
-        }
+    private fun redactLiteral(literal: PartiqlAst.Expr.Lit) {
+        val sourceLocation = literal.metas[SourceLocationMeta.TAG] as SourceLocationMeta?
+            ?: throw NoSuchElementException("No SourceLocation meta data in Literal object $literal")
```

Comment:
There's an extension property `sourceLocation` which should be used instead, also that doesn't seem like the right exception because [`NoSuchElementException`](https://docs.oracle.com/javase/8/docs/api/java/util/NoSuchElementException.html) is commonly thrown by `Iterator.next()` when there are no remaining elements.  I think a simple call to `error` should suffice here, and the other location I see pattern.

```
val sourceLocation = literal.metas.sourceLocation ?: error("Cannot redact due to missing source location")

Reply:
I agree, the error should be changed. The extension property `sourceLocation` was only for the `partiql.lang.ast.MetaContainer`, so I also added an equivalent one for IonElement MetaContainer in `SourceLocationMeta.kt`.